sdk/python/kfp/dsl/_component_bridge.py

```diff
@@ -124,4 +142,251 @@ def _create_container_op_from_component_and_arguments(
+        raise NotImplementedError(
+            'Currently, a component input can only accept either a constant '
+            'value or a reference to another pipeline parameter. It cannot be a '
+            'combination of both. Got: {} for input {}'.format(
+                argument_value, input_name))
+
+      input_type = component_spec._inputs_dict[input_name].type
+      if type_utils.is_parameter_type(input_type):
+        pipeline_task_spec.inputs.parameters[
+            input_name].runtime_value.constant_value.string_value = (
+                argument_value)
+      else:
+        # An importer node with constant value artifact_uri will be inserted.
+        pipeline_task_spec.inputs.artifacts[
+            input_name].task_output_artifact.producer_task = ''
+        type_schema = type_utils.get_input_artifact_type_schema(
+            input_name, component_spec.inputs)
+        importer_specs[input_name] = importer_node.build_importer_spec(
+            input_type_schema=type_schema, constant_value=argument_value)
+    elif isinstance(argument_value, int):
+      pipeline_task_spec.inputs.parameters[
+          input_name].runtime_value.constant_value.int_value = argument_value
+    elif isinstance(argument_value, float):
+      pipeline_task_spec.inputs.parameters[
+          input_name].runtime_value.constant_value.double_value = argument_value
+    elif isinstance(argument_value, _container_op.ContainerOp):
+      raise TypeError(
+          'ContainerOp object {} was passed to component as an input argument. '
+          'Pass a single output instead.'.format(input_name))
+    else:
+      if is_compiling_for_v2:
+        raise NotImplementedError(
+            'Input argument supports only the following types: PipelineParam'
+            ', str, int, float. Got: "{}".'.format(argument_value))
+
+  if not component_spec.name:
+    component_spec.name = _components._default_component_name
+
+  # task.name is unique at this point.
+  pipeline_task_spec.task_info.name = (dsl_utils.sanitize_task_name(task.name))
+  pipeline_task_spec.component_ref.name = (
+      dsl_utils.sanitize_component_name(component_spec.name))
+
+  task.task_spec = pipeline_task_spec
+  task.importer_specs = importer_specs
+  task.component_spec = dsl_component_spec.build_component_spec_from_structure(
+      component_spec)
+
+  resolved_cmd = _resolve_commands_and_args_v2(
+      component_spec=component_spec, arguments=original_arguments)
+
+  task.container_spec = (
+      pipeline_spec_pb2.PipelineDeploymentConfig.PipelineContainerSpec(
+          image=component_spec.implementation.container.image,
+          command=resolved_cmd.command,
+          args=resolved_cmd.args))
+
+  # Override command and arguments if compiling to v2.
+  if is_compiling_for_v2:
+    task.command = resolved_cmd.command
+    task.arguments = resolved_cmd.args
+
+    # limit this to v2 compiling only to avoid possible behavior change in v1.
+    task.inputs = input_params
```

Comment:
@neuromage, I will attempt to remove the above `if is_component_for_v2:` branch in a separate PR. There're some weird issues if I remove this, yet I haven't figured out the best way to fix it. So just keep it for now, so it reserves the current behaviors for both v1 and v2 compiling.

TL;DR
The symptom of removing this branch is when I compile the custom job component (https://github.com/kubeflow/pipelines/pull/5146), `task.command`, if not override by the v2 resolved command, will contain `{{pipelineparam:op=;name=pipeline-output-directory}}` as a result of https://github.com/kubeflow/pipelines/pull/5122.

and `task.inputs` is empty, so code below 
(specifically `self._inputs += _pipeline_param.extract_pipelineparams_from_any(...)`) will recognize there's a `PipelineParam` from the command and add it to the inputs. 
https://github.com/kubeflow/pipelines/blob/965bc62fd60e94de9d26603c1c06e60d931906db/sdk/python/kfp/dsl/_container_op.py#L769-L785

Having this extra input (`pipeline-output-directory`) then results an additional input in `component_spec.input_definitions`, which is wrong for v2 compilation - we don't need the pipeline root appear in the pipeline input in v2.

Reply:
I see, ok, thanks for explaining. We'll fix in follow up PR then.